### DISCUSSION TITLE
Shopware

### DISCUSSION THREAD

**Discussion by andrasbacsai** [coollabsio/coolify#3304] [Service Template Requests]
https://github.com/shopware/shopware

Docker image: https://docs.dockware.io/

**Comment by vanWittlaer** [comment]
Not sure whether dockware is the way to go here (for production). You may want to consider https://github.com/shopware/docker instead.

**Comment by BenRoe** [comment]
Can someone explain how to get sw6 running on coolify?

**Reply by ZenelShabani** [reply]
Best is to use the shopware/docker plugin (not dockware for production) and just deploy a docker project on coolify (this is how i do it for production) https://github.com/shopware/docker ( u must expose port 8000 i think is default in coolify settings page) also add this to post-deployment command : php bin/console system:update:finish && php bin/console theme:compile

**Comment by vanWittlaer** [comment]
As @ZenelShabani writes, you should start with the Shopware Docker image. You will find a sample docker-compose.yaml in the docs that will provide some initial guidance.

The point with Shopware is that, in addition to the web server, you need a scheduler and a worker component. With Coolify, these would be created as separate applications, as you have probably already realised. You may want to take a look at https://github.com/vanWittlaer/swoofy for more details, such as a sample Docker Compose YAML file and a graceful-shutdown script for the worker. I'm looking forward to receiving your feedback.

**Reply by BenRoe** [reply]
Already found swoofy, but it's still so complicated. A step by step guide would be awesome or a coolify service template.

**Reply by Cinzya** [reply]
I don't think we can make it a Coolify Service Template, as you probably want to connect it with your own repo, but we could provide a step by step guide on how to deploy, like we are already doing with other frameworks.

I will take a look into this and include it officially to the docs, @vanWittlaer probably copying a lot from your setup, seems pretty solid. 😄
Shouldn't hopefully be too difficult as I work with Shopware at my day job already.

**Reply by BenRoe** [reply]
Hello,
thank you for your reply. 
Why not a template to start from scratch like this command `shopware-cli project create .`?
If you can do a step by step, this would be very helpful. 🙏

**Reply by Cinzya** [reply]
I guess that depends on the usecase. If you never want to touch the underlying codebase and only use themes and extensions from the Shopware marketplace, a template will work.
But as soon as you want to customize Shopware yourself without an extension, you will need to touch the code. Which would then require it being deployed as an application and not service.

Could definitely offer both.

**Reply by BenRoe** [reply]
That sounds great. For example i start with a coolify template, because right now it's easiest and need only extension from marketplace/templates. Later i realise i need to customise the codebase. I could export db and existing codebase?
Could you estimate how long it will take to create a template?

**Reply by BenRoe** [reply]
@Cinzya I tried to get Shopware running in coolify with the https://github.com/vanWittlaer/swoofy repository, but can't get i running.
Looking forward for your shopware coolify template. 🥺

**Reply by BenRoe** [reply]
Here is my code. It's WIP. https://github.com/BenRoe/docker-compose-shopware6/tree/no-ci
Problem is to create the Shopware files inside the `/var/www/html` container.

**Reply by vanWittlaer** [reply]
> @Cinzya I tried to get Shopware running in coolify with the https://github.com/vanWittlaer/swoofy repository, but can't get i running. Looking forward for your shopware coolify template. 🥺

Hi @BenRoe swoofy is a fully functioning sample setup that powers https://swoofy.my-demo.shop/. Could you help us improve it by telling us exactly what the issue is that's preventing you from getting your setup running?

**Reply by BenRoe** [reply]
@vanWittlaer i opened a issue https://github.com/vanWittlaer/swoofy/issues/2

Please provide more information about how to get the local project to coolify.
Do you have a single compose with all necessary containers (redis, MariaDB, RabbitMQ, worker, scheduler), or do you run each of these separate in coolify?
It would be also nice if swoofy can be started without Shopware files in the swoofy/shopware folder. Because i have to check, or trust the code inside.

**Reply by vanWittlaer** [reply]
> @vanWittlaer i opened a issue [vanWittlaer/swoofy#2](https://github.com/vanWittlaer/swoofy/issues/2)
> 
> Please provide more information about how to get the local project to coolify. Do you have a single compose with all necessary containers (redis, MariaDB, RabbitMQ, worker, scheduler), or do you run each of these separate in coolify? It would be also nice if swoofy can be started without Shopware files in the swoofy/shopware folder. Because i have to check, or trust the code inside.

TL;DR: Shopware with Coolify must be set up as a project; it's not just an app or service. In other words, run the web server, database, Redis, workers, etc. as separate services.

I don't fully understand your point about the shopware/docker folder. This is where Shopware and your plugin sources reside. If you leave this empty, you will just have a basic web server.

**Comment by Cinzya** [comment]
Just added a example for Shopware 6 to our [official Coolify examples](https://github.com/coollabsio/coolify-examples/tree/v4.x/shopware6).

This was based on [Shopwares own example repository](https://github.com/shopwareLabs/example-docker-repository), with slight changes to support Coolify envs and a few missing dependencies for Shopware 6.7. This wasn't tested yet against every scenario, so might need some further changes, but from my initial test, it builds and runs fine to me in a [v4.0.0-beta.453](https://github.com/coollabsio/coolify/releases/tag/v4.0.0-beta.453) coolify instance.

Let us know if you have any suggestions on how we can improve it or if you face any errors.

**Reply by BenRoe** [reply]
Thank you for the example.
Here my questions, for better understanding:
1. Isn't it better to use a env variable In the [MariaDB healthcheck ](https://github.com/coollabsio/coolify-examples/blob/be1230852691bcd849c17da6b7c857629452d742/shopware6/docker-compose.yaml#L56), instead of a hardcoded password?
2. You use Tmpfs instead of a volume for the database. After restarting the server, all data are gone? Can you explain why you use Tmpfs?
3. Why valkey instead of redis?

**Reply by BenRoe** [reply]
Deploying was successful. Open the storefront shows "Sales Channel Not Found". That's fine and tried to open https://url.com/admin but the site is white.
Error in the browser console
```
[Warning] [blocked] The page at https://web-zcksokg4o0.c.mydoamin.com/admin requested insecure content from https://web-zcksokg4o0.c.mydoamin.com/bundles/administration/administration/assets/administration-BrQIsQjg.js. This content was blocked and must be served over HTTPS.
```

**Reply by Cinzya** [reply]
No specific reason for those, it's what I copied / is left over from Shopwares own example compose, but we can adjust. 
Not sure why they use / recommend valkey, but if you do some research online, there are claims that valkey performs better.

I got the admin error only on pre-load, after 5 seconds or so it won't be white anymore. But I tested against a pre-generated `sslip.io` domain on `http` might behave different if you try to serve over `https` can take a look at that.

**Reply by BenRoe** [reply]
@Cinzya fixed the http problem and white page with
```
# config/packages/trusted_env.yaml
parameters:
  env(TRUSTED_PROXIES): ''

framework:
  trusted_proxies: '%env(TRUSTED_PROXIES)%'
  trusted_headers:
    - 'x-forwarded-for'
    - 'x-forwarded-host'
    - 'x-forwarded-proto'
    - 'x-forwarded-port'
    - 'x-forwarded-prefix'
```
also had to add this env var

```
TRUSTED_PROXIES: '${TRUSTED_PROXIES:-172.16.0.0/12,10.0.0.0/8}'
```

**Reply by BenRoe** [reply]
Did a research about valkey as alternative to redis.
  | Category       | Result                                              |
  |----------------|-----------------------------------------------------|
  | Compatibility  | Officially recommended by Shopware 6.7           |
  | Performance    | Up to 28% less memory, better multi-core scaling |
  | Security       | Equivalent to Redis (ACL, TLS, auth)             |
  | Resource Usage | Lower memory footprint than Redis                |
  | Licensing      | BSD 3-Clause (no restrictions)                   |
